Added/filled admin configuration page

sirzento opened the pull request:
I filled the missing part from "admin configuration".
I hope that this is the right content for this section.

Preview:

![grafik](https://github.com/user-attachments/assets/7352e929-2f8c-4c07-9626-bf55cae59a6b)

prplwtf commented on docs/pages/developing-extensions/Admin-configuration.md:
```diff
@@ -1,21 +1,213 @@
+# Admin configuration
+<h4 class="fw-light">Add configuration options to your extension's admin page.</h4><br/>
+
+This example demonstrates how to create an admin controller that loads and saves extension-specific configuration values using the `BlueprintExtensionLibrary` database utility. It also includes input validation via a custom form request.
```
When mentioning BlueprintExtensionLibrary in documentation, hyperlink it to the $blueprint documentation,

prplwtf commented on docs/pages/developing-extensions/Admin-configuration.md:
```diff
@@ -1,21 +1,213 @@
+- `AdminFormRequest`: Handles form input validation.
+- `RedirectResponse`: Used to return the user back to the admin page after saving changes.
+
+Inject the `SettingsRepositoryInterface` into your controller constructor:
```
"controller constructor" should be "view constructor"

sirzento replied:
But isn't this the constructor of the controller?

prplwtf replied:
Whoops!

prplwtf commented on docs/pages/developing-extensions/Admin-configuration.md:
````diff
@@ -1,21 +1,213 @@
+
+### Loading configuration and setting defaults
+
+To handle configuration, extend your `index()` function to retrieve the current values (or fall back to defaults if none exist). 
+
+For this example, we will add the configuration `theme` and `customName`:
+
+```php
+public function index(): View
+{
+    $theme = $this->blueprint->dbGet('{identifier}', 'theme');
+    $customName = $this->blueprint->dbGet('{identifier}', 'customName');
+
+    $defaulttheme = 0;
+    $defaultCustomName = "superuser";
+
+    if ($theme == "") {
+        $this->blueprint->dbSet('{identifier}', 'theme', $defaulttheme);
+        $theme = $defaulttheme;
+    }
+    if ($customName == "") {
+        $this->blueprint->dbSet('{identifier}', 'customName', $defaultCustomName;
+        $customName = $defaultCustomName;
+    }
+
+    return $this->view->make(
+      'admin.extensions.{identifier}.index', [
+        'theme' => $theme,
+        'customName' => $customName,
+        'root' => "/admin/extensions/{identifier}",
+        'blueprint' => $this->blueprint,
+    ]
+);
+
+```
````
This is common behavior across extensions, but is not the proper way to do it. This should be done through migrations instead.

prplwtf replied:
I'm not quite sure how you'd explain this correctly though. Easier to add a note there that this is an oversimplified explanation and while functional, developers should use migrations instead.

sirzento replied:
I thought this would be the right way to store extension wide settings. Thats even how the "admin configuration" template does it.

prplwtf replied:
Yeahhh.. about that..

This was how I did it in the past, and it was meant to be a band-aid fix for my stupidity. While it _works_ fine, it's not how developers should do it because it can be awful for updating extensions.

For people just starting extension development it's fine though, we just also need to push people into the direction of migrations, as they're better.

Best you can do is keep this in, but also tell users that it's not great when shipping updates and migrations are recommended over this.

sirzento replied:
But if this isn't the way to go, why even mention/show it at all?

I think I need to rewrite it to use logic described in the "Custom table and migrations" part. I would also move this part after "Custom table and migrations" part but be able to mention as "part what you already know/have done".

That makes me wonder what the intended use case is for `dbGet()` and `dbSet()` if it shouldn't be used for extension settings.

prplwtf replied:
dbGet() can be used for getting records, dbSet() can be used for setting records.

The initial record creation should be done through migrations.

Blueprint's db functions are generally made to allow a more simple and straightforward way to interact with the database. Nebula, for example, uses dbGet() inside of Designer to get records and in the dashboard wrapper to get it's configuration.

Blueprint's db functions also give us more control over how developers interact with the database.

sirzento replied:
I got everything done now but I am still unsure about this one. 

> The initial record creation should be done through migrations.

Would something like this work? I can't test right now since I somehow bricket my installation while testing :joy:

```php
use Illuminate\Support\Facades\DB;
use Illuminate\Database\Migrations\Migration;

class AddDefaultExtensionSettings extends Migration
{
    public function up()
    {
        DB::table('settings')->insert([
            'key' => '{identifier}::theme',
            'value' => serialize('0'),
        ],
        [
            'key' => '{identifier}::customName',
            'value' => serialize('superuser'),
        ],
        [
            'key' => '{identifier}::count',
            'value' => 10,
        ]);
    }

    public function down()
    {
        DB::table('settings')->where('key', 'like', '{identifier}::%')->delete();
    }
}
```

prplwtf replied:
The class name can conflict with other extensions, instruct users to name their file something unique to their extension. Besides that, I _think_ the migration looks good.

prplwtf commented on docs/pages/developing-extensions/Admin-configuration.md:
````diff
@@ -1,21 +1,213 @@
+
+```
+
+The `$blueprint->dbGet()` function is used to retrieve values from the database. If no value is found, the default is applied and saved using dbSet().
````
```diff
- dbSet()
+ `dbSet()`
```

prplwtf commented on docs/pages/developing-extensions/Admin-configuration.md:
````diff
@@ -1,21 +1,213 @@
+
+Finally, create a form request class at the bottom of the controller file (outside of your controller class) to validate incoming form data:
+
+```php
+class {identifier}SettingsFormRequest extends AdminFormRequest
+{
+    public function rules(): array
+    {
+        return [
+            'theme' => 'string',
+            'customName' => 'string',
+        ];
+    }
+
+    public function attributes(): array
+    {
+        return [
+            'theme' => 'Theme',
+            'customName' => 'Custom Name,
+        ];
+    }
+}
+
+```
````
Maybe good to document other kind of option possibilities that aren't just strings. 3 options total should be fine, maybe one number between 1 and 10, a string and a string that has to pass a regex rule.

For simplicity, instead of separating rules with pipes `|`, store them in an array:
```diff
- 'string|in:foo,bar'
+ ['string', 'in:foo,bar']
```

prplwtf commented on docs/pages/developing-extensions/Admin-configuration.md:
```diff
@@ -1,21 +1,213 @@
+    placeholder="type here"
+    class="form-control"
+  />
+  <select class="form-control" name="theme">
+    <option value="0" @if($theme == "0") selected @endif>Theme 1</option>
+    <option value="1" @if($theme == "1") selected @endif>Theme 2</option>
+    <option value="2" @if($theme == "2") selected @endif>Theme 3</option>
+  </select>
```
This seems to be an option select while the validation rules accept all strings.

sirzento replied:
The "admin configuration" template does also handle this input as string but I have the feeling that the template isn't a reliable source anymore haha 

Would it be alright if I just change the input validation to `['string', 'in:0,1,2']` ?

prplwtf replied:
You can do that or, I think, you can handle it as a number with a gte:0 and lte:2.

gte: greater than equal, number must match gte value or be higher
lte: less than equal, number must match lte value or be lower

I can double check later

sirzento commented:
Alright, I added a part that explains that migration are needed and removed the code that sets the defaults via inserts. Would it be alright like this?

Preview:

![grafik](https://github.com/user-attachments/assets/3f861ba2-7506-4ee2-a097-2517069624b1)

0x7d8 commented on docs/pages/developing-extensions/Admin-configuration.md:
```diff
@@ -1,21 +1,262 @@
+public function update({identifier}SettingsFormRequest $request): RedirectResponse
+{
+    foreach ($request->normalize() as $key => $value) {
+        $this->settings->set('{identifier}::' . $key, $value);
```
```php
foreach ($request->normalize() as $key => $value) {
  $this->blueprint->dbSet("{identifier}", $key, $value);
}
```

or

```php
$this->blueprint->dbSetMany("{identifier}", $request->normalize());
```

0x7d8 commented on docs/pages/developing-extensions/Admin-configuration.md:
````diff
@@ -1,21 +1,262 @@
+
+[Create a migration file](?page=developing-extensions/Custom-table-and-migrations) and include the following logic:
+
+```php
````
```php
<?php

use Illuminate\Support\Facades\DB;
use Illuminate\Database\Migrations\Migration;
use Pterodactyl\BlueprintFramework\Libraries\ExtensionLibrary\Admin\BlueprintAdminLibrary as BlueprintExtensionLibrary;

return new class extends Migration
{
    public function up(): void
    {
        $blueprint = app(BlueprintExtensionLibrary::class);

        $blueprint->dbSetMany("{identifier}", [
            'theme' => '0',
            'customName' => 'superuser',
            'count' => 10,
        ]);
    }

    public function down(): void
    {
        DB::table('settings')->where('key', 'like', '{identifier}::%')->delete();
    }
};
```

0x7d8 reviewed (approved):
will work, some suggestions though

sirzento commented:
Alright I think everything should be fine now. Using the `BlueprintExtensionLibrary` inside the migration really does shorten the code a bit and normalize the everything a bit more to always use the `BlueprintExtensionLibrary` for database handling.

ab00ky reviewed (changes requested):
How can I use this to to identify my device admin

prplwtf reviewed (approved):
LGTM, is this ready to be merged?

sirzento commented:
From my side yes
